`WCProductSettingsModel` to `Room`

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: AINFRA-536

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR migrates `WCProductSettingsModel` to `Room`

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

1. On the web, go to the store settings
2. Change the products settings (`WooCommerce` > `Settings` > `Products` > `Weight unit`)
3. Run the mobile app: assert the new unit is set (go to product creation, add more details, shipping)
4. Change the unit settings again
5. Restart the mobile app: assert the unit is updated as well

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="1040" alt="image" src="https://github.com/user-attachments/assets/8f63d3fb-a2c6-48b8-8d3b-d0f76f8a29fa" />


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate `WCProductSettingsModel` to be a Room entity

wzieba committed:
Adjust WooCommerce module codebase to match new WCProductSettingsModel entity

wzieba committed:
Adjust unit tests to updated `WCProductSettingsModel`

Also, moved more test utilities to `testFixtures`.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ProductSettingsDao</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Revert import changes on `WooCommerceStoreTest`

To reduce diff and improve readability

wzieba committed:
Rename Room entity to ProductSettingsEntity

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductSettingsModel.kt:
```diff
@@ -1,21 +1,14 @@
-        this.id = id
-    }
-}
+    @PrimaryKey val localSiteId: Int = 0,
```
Suggestion (💡): Just like done in all the previous migrations, consider refactor this entity to utilize `LocalId` in favor of `Int`.

ParaskP7 reviewed (changes requested):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected (ish), another migration (almost) done, with no more `WCProduct.*` prefixed tables left within `wp-fluxc`, nice! ✅ 💪 🚀 

---

Warning (⚠️): This ish above is just because of the fact that during this migration I am not seeing the `WCProductSettingsModel` table being dropped from the `wp-fluxc` database (just like it was done [here](https://github.com/woocommerce/woocommerce-android/pull/14027/commits/4211e37dc37742a6c59d3da7f22a9c62f921f740) as an example).

`WellSqlConfig.kt`
```kotlin
open class WellSqlConfig : DefaultWellConfig {
    ...
    override fun getDbVersion(): Int {
        return 215
    }

    override fun onUpgrade(...) {
                ...
                214 -> migrateAddOn(ADDON_WOOCOMMERCE, version) {
                    db.execSQL("DROP TABLE IF EXISTS WCProductSettingsModel")
                }
                ...
    }
    ...
}
```

-----

I have left one warning (⚠️) and one suggestions (💡) for you to consider. I am **NOT** going to approve this PR for now, just because of this one warning, instead I am hitting request changes.

---

PS: I am not sure if you also want someone from the product team to do some more testing on this change before merging. I'll leave that up to you. 👍

wzieba committed:
Use `LocalId` for site id in WCProductSettingsModel

wzieba committed:
Add migration to remove old `WCProductSettingsModel` from WellSql database

wzieba commented:
Thanks for the review @ParaskP7 , I forgot about the migration again. The more I think about it, I think we maybe shouldn't delete tables one by one and create dozens of migration, but rather once at the end of the project?

Anyway, I added the migration and rebased on top of `db/remove-unused-tables-from-every-database` to prematurely address the merge conflict with your PR.

ParaskP7 commented:
Awesome, thanks for the update @wzieba , everything LGTM now! ✅ 

> The more I think about it, I think we maybe shouldn't delete tables one by one and create dozens of migration, but rather once at the end of the project?

To be honest, I am not sure, I think us progressively migrating and dropping the tables as we go seems more prudent, especially since us migrating away from every table might take us some time, probably months. And anyway, it doesn't hurt since this whole `WellSQL` configuration file will be removed at the end. So, no matter this `getDbVersion()` and `onUpgrade(...)` version numbers could get as high as needed, until totally removed... 🤷 

> Anyway, I added the migration and rebased on top of db/remove-unused-tables-from-every-database to prematurely address the merge conflict with your PR.

Oh cool, nice, I'll merge my PR and then this on Monday and we'll move forward... 🚀

ParaskP7 commented:
> Oh cool, nice, I'll merge my PR and then this on Monday and we'll move forward... 🚀

Ah, actually you can merge this on my branch and I can do the merging of mine on Monday, up to you. 👍

ParaskP7 commented:
FYI: @wzieba I'll now start the merging process, this PR first, then #14107 after I update the latest `trunk` on it.

wzieba commented:
I think for the future we shouldn't do it like this.

Now, when I go to #14107 I can't clearly see what code changes are related to removing unused tables, and what work is about migrating `WCProductSettingsModel` :(

For the next time, please merge the PR that targets trunk (#14107) first, and then the PR that was targeting it (#14096) as it will be automatically updated to target trunk.

ParaskP7 commented:
> I think for the future we shouldn't do it like this.

I agree, but actually, I would go about saying that we should each merge their own PRs into `trunk`, especially when those are (totally) unrelated, that is, without depending on each other, even if that means that the other one would need to apply some merge conflicts or increase the database migration number, etc. 🤷

> Now, when I go to https://github.com/woocommerce/woocommerce-android/pull/14107 I can't clearly see what code changes are related to removing unused tables, and what work is about migrating WCProductSettingsModel :(
> 
> For the next time, please merge the PR that targets trunk (https://github.com/woocommerce/woocommerce-android/pull/14107) first, and then the PR that was targeting it (https://github.com/woocommerce/woocommerce-android/pull/14096) as it will be automatically updated to target trunk.

True, I just didn't think about us wanting to go back to this #14107 PR and traverse the changes! 👍

wzieba commented:
> we should each merge their own PRs into trunk

I agree and if this PR wouldn't be merged to #14107 , it'd be exactly like this. This is why I added "do not merge" label 👍.

> True, I just didn't think about us wanting to go back to this https://github.com/woocommerce/woocommerce-android/pull/14107 PR and traverse the changes

It's less about examining changes now, but rather in the future.

ParaskP7 commented:
> I agree and if this PR wouldn't be merged to https://github.com/woocommerce/woocommerce-android/pull/14107 , it'd be exactly like this. This is why I added "do not merge" label 👍.

Yea, I totally missed your intention here, to have this PR ready for when #14107 gets merged, you wanting to be a step ahead so that you don't have to do any work afterwords... 🫣

> It's less about examining changes now, but rather in the future.

Yea, I meant when in the future too... 👍
